x/oracle/keeper/abci.go

```diff
@@ -6,164 +6,197 @@ import (
+				// setting it to store.
+				panic(err)
 			}
+
+			// add oracle data to maps
+			submittedFeedMap[oracleData.Type()][validatorAddr.String()] = oracleData
+
 		}
 
-		// if the voted_power/total_power > params.VoteThreshold then we store the averages in the store
-		storeAverages := sdk.NewDec(votedPower).Quo(sdk.NewDec(totalPower)).GT(params.VoteThreshold)
+		// delete the vote as it has already been processed
+		// TODO: consider keeping the votes for a few voting windows in order to
+		// be able to submit evidence of inaccurate data feed
+		k.DeleteOracleDataVote(ctx, validatorAddr)
+		return false
+	})
+
+	// iterate over the full list of validators to increment miss counters if they didn't vote
+	totalPower := int64(0)
+	k.stakingKeeper.IterateBondedValidatorsByPower(ctx, func(_ int64, validator stakingtypes.ValidatorI) bool {
+		totalPower += validator.GetConsensusPower()
+		validatorAddr := sdk.AccAddress(validator.GetOperator())
+
+		if !validatorVotesMap[validatorAddr.String()] {
+			// TODO: this is wrong because the feeder could submit a single oracle data uniswap pair instead of all the required
+			// ones and still be counted as if they voted correctly. Maybe consider adding each uniswap pair id to the params?
+			// TODO: we need to define what is the exact data that we want the validators to submit.
+			k.IncrementMissCounter(ctx, validatorAddr)
+		}
 
-		// compute the averages for each type of data tracked by the oracle
-		for typ, dataCollection := range collectionMap {
-			// first, lets delete the old data
-			k.DeleteOracleData(ctx, typ)
+		return false
+	})
 
-			// then we compute the "average"
-			avg := types.GetAverageFunction(typ)(dataCollection)
+	// if the voted_power/total_power > params.VoteThreshold then we store the averages in the store
+	// TODO: check for sample size too as there could be only a few feeders with a large voting power that submitted the
+	// data to the oracle.
+	storeAverages := sdk.NewDec(votedPower).Quo(sdk.NewDec(totalPower)).GT(params.VoteThreshold)
 
-			// once we have an "average" we set it in the store
-			if storeAverages {
-				k.SetOracleData(ctx, avg)
-			}
+	// Now, compute the aggregated data (eg: avg, median, etc.) for each type of data tracked by the oracle
 
-			// store the "average" for scoring validators later
-			averageMap[typ] = avg
+	// NOTE: we iterate over the params data types to avoid using map iteration
+	// which is non-deterministic.
+	for _, dt := range params.DataTypes {
+		oracleData, ok := oracleDataByTypeMap[dt]
+		if !ok {
+			// no oracle data type was submitted
+			continue
 		}
 
-		// Compare each validators vote for each data type against the
-		// averages to define which are eligable for rewards
-		for dataType, vals := range detailedMap {
-			for val, data := range vals {
-				rewardEligable[val] = false
-				if averageMap[dataType].Valid(data) {
-					rewardEligable[val] = true
-				}
-			}
-		}
+		// TODO: delete the oracle old data
```

Comment:
Why is this still todo? Do we need to decide how many periods of data we are saving?